File: bindings/swift/Sources/Examples/JsonGrammar.swift
import Foundation
import XGrammar

/// Compile the built-in JSON grammar and accept a sample JSON string.
public func runJsonGrammar() throws {
    // Minimal byte-level vocab: one token per byte 0..=255, plus a stop token.
    var encodedVocab: [String] = (0..<256).map { byte in
        String(bytes: [UInt8(byte)], encoding: .isoLatin1)!
    }
    encodedVocab.append("</s>")

    let tokenizerInfo = try TokenizerInfo(
        encodedVocab: encodedVocab,
        vocabType: 1, // VocabType::BYTE_FALLBACK
        vocabSize: nil,
        stopTokenIds: [256],
        addPrefixSpace: false
    )
    let compiler = GrammarCompiler(
        tokenizerInfo: tokenizerInfo,
        maxThreads: 8,
        cacheEnabled: true,
        cacheLimitBytes: -1
    )
    let compiled = compiler.compileBuiltinJsonGrammar()
    let matcher = GrammarMatcher(
        compiledGrammar: compiled,
        overrideStopTokens: nil,
        terminateWithoutStopToken: false,
        maxRollbackTokens: -1
    )

    let sample = #"{"name":"Ada","role":"student"}"#
    precondition(matcher.acceptString(input: sample, debugPrint: false), "grammar rejected: \(sample)")
    print(sample)
    print("completed=\(matcher.isCompleted()) terminated=\(matcher.isTerminated())")
}

File: crates/xgrammar/examples/json_grammar.rs
//! Compile the built-in JSON grammar and accept a sample JSON string.

use xgrammar::{GrammarCompiler, GrammarMatcher, TokenizerInfo, VocabType};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Minimal byte-level vocab: one token per byte 0..=255, plus a stop token.
    let mut encoded_vocab: Vec<String> = (0u8..=255)
        .map(|byte| char::from_u32(u32::from(byte)).unwrap().to_string())
        .collect();
    encoded_vocab.push("</s>".to_string());

    let tokenizer_info = TokenizerInfo::new(
        &encoded_vocab,
        VocabType::ByteFallback,
        None,
        Some(vec![256]),
        false,
    );
    let compiler = GrammarCompiler::with_defaults(tokenizer_info);
    let compiled = compiler.compile_builtin_json_grammar();
    let mut matcher = GrammarMatcher::from_compiled_grammar(&compiled, false);

    let sample = r#"{"name":"Ada","role":"student"}"#;
    assert!(
        matcher.accept_string(sample),
        "grammar rejected: {sample}"
    );
    println!("{sample}");
    println!(
        "completed={} terminated={}",
        matcher.is_completed(),
        matcher.is_terminated()
    );
    Ok(())
}
